custom/icds_reports/management/commands/dashboard_usage_report.py

```diff
@@ -0,0 +1,156 @@
+                _write_to_file(cache_name, data)
+            return data
+        return _inner
+    return _outer
+
+
+def _get_from_file(filename):
+    if os.path.exists(filename):
+        print(f'Fetching data from file: {filename}')
+        with open(filename, 'r') as f:
+            reader = csv.reader(f)
+            return [
+                r[0] if len(r) == 1 else r for r in list(reader)
+            ]
+
+
+def _write_to_file(filename, rows):
+    print(f'Writing {len(rows)} to file {filename}')
+    with open(filename, 'w') as f:
+        writer = csv.writer(f)
+        writer.writerows([
+            r if isinstance(r, list) else [r] for r in rows
+        ])
+
+
+class Command(BaseCommand):
+    def add_arguments(self, parser):
+        parser.add_argument(
+            'start_date',
+            type=date_type,
+            help='The start date (inclusive). format YYYY-MM-DD'
+        )
+        parser.add_argument(
+            'end_date',
+            type=date_type,
+            help='The end date (exclusive). format YYYY-MM-DD'
+        )
+        parser.add_argument('domain')
+        parser.add_argument('username', help='The username of the logged in user')
+
+    def handle(self, start_date, end_date, domain, username, **options):
+        print(f'fetching users data')
+
+        user = CouchUser.get_by_username(username)
+        usage_data = DashBoardUsage(couch_user=user, domain=domain).get_excel_data()[0][1]
+
+        # fetching dashboard usage counts
+        tab_usage_counts, tab_indicators_count, cas_counts = \
+            self.get_dashboard_usage_counts(start_date, end_date, domain)
+
+        username_state_mapping = self.get_tabular_data(usage_data, tab_usage_counts, tab_indicators_count)
+        print(f'Request data written to file {REQUEST_DATA_CACHE}')
+        self.get_cas_data(cas_counts, username_state_mapping)
+        print(f'Request data written to file {CAS_DATA_CACHE}')
+
+    def get_dashboard_usage_counts(self, start_date, end_date, domain):
+        """
+        :param start_date: start date of the filter
+        :param end_date: end date of the filter
+        :param domain
+        :return: returns the counts of no of downloads of each and total reports for  all usernames
+        """
+        print(f'Compiling usage counts for users')
+        tabular_user_counts = defaultdict(int)
+        cas_user_counts = defaultdict(int)
+        tabular_user_indicators = defaultdict(lambda: [0] * 10)
+
+        urls = ['/a/' + domain + '/cas_export', '/a/' + domain + '/icds_export_indicator']
+        query = list(
+                ICDSAuditEntryRecord.objects.filter(url__in=urls, time_of_use__gte=start_date,
+                                                    time_of_use__lte=end_date)
+                .annotate(indicator=Cast(KeyTextTransform('indicator', 'post_data'), IntegerField()))
+                .filter(indicator__lte=THR_REPORT_EXPORT).values('indicator', 'username', 'url')
```

Comment:
This might fail `indicator__lte=THR_REPORT_EXPORT` because in cas data export indicators are not numbers they are alpha strings.

Reply:
It would be good to just separate cas data export usage and tabular report usage into two different functions, When I said in last review I actually meant that.
In one fuction you can just query Icds Audit to find usage of Cas data export and in second function you can just query Icds audit to find usage of tabular report. 
This way they will be totally decoupled an failue of one wont cause failure of other.

Reply:
aah..okay. since we are getting results in a single query, i put them in a single fuction

Reply:
> This might fail `indicator__lte=THR_REPORT_EXPORT` because in cas data export indicators are not numbers they are alpha strings.

We are casting it to integer and then filtering it